src/java/org/jivesoftware/openfire/pubsub/PubSubEngine.java

```diff
@@ -1084,13 +1105,83 @@ private void getPublishedItems(PubSubService service, IQ iq, Element itemsElemen
     }
 
     private void createNode(PubSubService service, IQ iq, Element childElement, Element createElement) {
+        // Call createNodeHelper and get the node creation status.
+        CreateNodeResponse response = createNodeHelper(service, iq, childElement, createElement);
+        if(response.getNewNode() == null) {
+            // New node creation failed
+            sendErrorPacket(iq, response.getStatus(), response.getPubsubError());
+        } else {
+            IQ reply = IQ.createResultIQ(iq);
+            Node newNode = response.getNewNode();
+            String nodeID = createElement.attributeValue("node");
+            // Include new nodeID if it has changed from the original nodeID
+            if (!newNode.getNodeID().equals(nodeID)) {
+                Element elem = reply.setChildElement("pubsub", "http://jabber.org/protocol/pubsub");
+                elem.addElement("create").addAttribute("node", newNode.getNodeID());
+            }
+            router.route(reply);
+        }
+    }
+
+    /**
+     * Response Object returned by createNodeHelper method
+     */
+    private class CreateNodeResponse {
+        private PacketError.Condition creationStatus;
+        private Node newNode;
+        private Element pubsubError;
```

Comment:
I think these should be set within a constructor, and possibly should be final. I don't think these should have setter methods.